File: neatformcore/src/main/java/com/nerdstone/neatformcore/rules/RulesFactory.java
package com.nerdstone.neatformcore.rules;

import android.content.Context;

import com.nerdstone.neatformcore.domain.model.NFormViewDetails;

import org.jeasy.rules.api.Facts;
import org.jeasy.rules.api.Rule;
import org.jeasy.rules.api.RuleListener;
import org.jeasy.rules.api.Rules;
import org.jeasy.rules.core.DefaultRulesEngine;
import org.jeasy.rules.mvel.MVELRuleFactory;
import org.jeasy.rules.support.JsonRuleDefinitionReader;
import org.jeasy.rules.support.YamlRuleDefinitionReader;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

import io.reactivex.Completable;
import timber.log.Timber;

import static com.nerdstone.neatformcore.datasource.AssetFile.openFileAsset;

public class RulesFactory implements RuleListener {

    private static final String TAG = RulesFactory.class.getCanonicalName();
    private static RulesFactory rulesFactory;
    private Rules rules;
    private Facts facts;
    private DefaultRulesEngine rulesEngine;
    private String currentRule;
    private Map<String, Set<String>> subjectsRegistry;

    private RulesFactory() {
        initRuleFactory();
    }

    public static RulesFactory getInstance() {
        if (rulesFactory == null) {
            rulesFactory = new RulesFactory();
        }
        return rulesFactory;
    }

    public Completable readRulesFromFile(Context context, String filePath, RulesFileType rulesFileType) {
        return Completable.fromAction(() -> {
            if (rules == null) {
                MVELRuleFactory mvelRuleFactory;
                if (rulesFileType == RulesFileType.JSON) {
                    mvelRuleFactory = new MVELRuleFactory(new JsonRuleDefinitionReader());
                } else if (rulesFileType == RulesFileType.YAML) {
                    mvelRuleFactory = new MVELRuleFactory(new YamlRuleDefinitionReader());
                } else {
                    throw new Exception("Unsupported file type. The library only supports YAML and JSON standards at the moment");
                }

                rules = mvelRuleFactory.createRules(new BufferedReader(
                        new InputStreamReader(openFileAsset(context, filePath))));
            }
        });
    }

    private void updateFacts(NFormViewDetails viewDetails) {
        facts.put(viewDetails.getName(), viewDetails.getValue());
    }

    private void fireRules() {
        rulesEngine.fire(rules, facts);
        Timber.i("rule fired");
    }

    private void initRuleFactory() {
        facts = new Facts();
        rulesEngine = new DefaultRulesEngine();
        rulesEngine.registerRuleListener(this);
        subjectsRegistry = new HashMap<>();
    }

    @Override
    public boolean beforeEvaluate(Rule rule, Facts facts) {
        Timber.tag(TAG).d("Before evaluation of rule {%s}", rule);
        return rule.getName().startsWith(getCurrentRule());
    }

    @Override
    public void afterEvaluate(Rule rule, Facts facts, boolean evaluationResult) {
        if (evaluationResult) {
            Timber.tag(TAG).d("Cool {%s} evaluated as 'true'", rule.getName());
        }
    }

    @Override
    public void beforeExecute(Rule rule, Facts facts) {
        Timber.tag(TAG).d("Executing rule {%s}", rule);
    }

    @Override
    public void onSuccess(Rule rule, Facts facts) {
        Timber.tag(TAG).d("Rule {%s} fired successfully:", rule);
    }

    @Override
    public void onFailure(Rule rule, Facts facts, Exception exception) {
        Timber.tag(TAG).e(exception);
    }

    private String getCurrentRule() {
        return currentRule;
    }

    private void setCurrentRule(String currentRule) {
        this.currentRule = currentRule;
    }

    public void updateFactsAndExecuteRule(NFormViewDetails viewDetails) {
        rulesFactory.setCurrentRule(viewDetails.getName());
        rulesFactory.updateFacts(viewDetails);
        rulesFactory.fireRules();
    }

    public Map<String, Set<String>> getSubjectsRegistry() {
        return subjectsRegistry;
    }

    public enum RulesFileType {
        JSON, YAML
    }
}

File: neatformcore/src/main/java/com/nerdstone/neatformcore/rules/RulesFactory.kt
package com.nerdstone.neatformcore.rules

import android.content.Context
import com.nerdstone.neatformcore.datasource.AssetFile
import com.nerdstone.neatformcore.domain.model.NFormRule
import com.nerdstone.neatformcore.domain.model.NFormViewDetails
import com.nerdstone.neatformcore.domain.model.NFormViewProperty
import com.nerdstone.neatformcore.domain.view.RulesHandler
import com.nerdstone.neatformcore.utils.Constants
import com.nerdstone.neatformcore.utils.Utils
import org.jeasy.rules.api.Facts
import org.jeasy.rules.api.Rule
import org.jeasy.rules.api.RuleListener
import org.jeasy.rules.api.Rules
import org.jeasy.rules.core.DefaultRulesEngine
import org.jeasy.rules.mvel.MVELRuleFactory
import org.jeasy.rules.support.JsonRuleDefinitionReader
import org.jeasy.rules.support.YamlRuleDefinitionReader
import org.mvel2.CompileException
import timber.log.Timber
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

class RulesFactory private constructor() : RuleListener {

    private var facts: Facts = Facts()
    private var rulesEngine: DefaultRulesEngine = DefaultRulesEngine()
    private lateinit var currentViewDetails: NFormViewDetails
    private var executableRulesList: HashSet<Rule> = hashSetOf()
    var allRules: Rules? = null
    val rulesHandler: RulesHandler = NFormRulesHandler.INSTANCE
    val subjectsRegistry: HashMap<String, HashSet<NFormRule>> = hashMapOf()

    init {
        rulesEngine.registerRuleListener(this)
    }

    override fun beforeEvaluate(rule: Rule?, facts: Facts?): Boolean = true

    override fun onSuccess(rule: Rule?, facts: Facts?) = Timber.d("%s executed successfully", rule)

    override fun onFailure(rule: Rule?, facts: Facts?, exception: Exception?) =
        when (exception) {
            is CompileException -> Timber.e(exception.cause)
            else -> Timber.e(exception)
        }

    override fun beforeExecute(rule: Rule?, facts: Facts?) = Unit

    override fun afterEvaluate(rule: Rule?, facts: Facts?, evaluationResult: Boolean) =
        rulesHandler.updateSkipLogicFactAfterEvaluate(evaluationResult, rule, facts)

    fun readRulesFromFile(
        context: Context, filePath: String, rulesFileType: RulesFileType
    ) {
        if (allRules == null) {
            val mvelRuleFactory: MVELRuleFactory = when (rulesFileType) {
                RulesFileType.JSON -> MVELRuleFactory(JsonRuleDefinitionReader())
                RulesFileType.YAML -> MVELRuleFactory(YamlRuleDefinitionReader())
            }

            allRules = mvelRuleFactory.createRules(
                BufferedReader(
                    InputStreamReader(AssetFile.openFileAsset(context, filePath))
                )
            )
        }
    }

    private fun updateCurrentViewAndFacts(viewDetails: NFormViewDetails) {
        currentViewDetails = viewDetails
        facts.put(currentViewDetails.name, currentViewDetails.value)
    }

    private fun fireRules() {
        rulesEngine.fire(Rules(executableRulesList), facts)
        rulesHandler.handleSkipLogic(facts)
    }

    fun updateFactsAndExecuteRules(viewDetails: NFormViewDetails) {
        updateCurrentViewAndFacts(viewDetails)
        updateExecutableRules()
        fireRules()
    }

    private fun updateExecutableRules() {
        executableRulesList.clear()
        val dependantViews = subjectsRegistry[currentViewDetails.name]
        dependantViews?.forEach { nFormRule ->
            nFormRule.matchingRules.also {
                if (it.isEmpty()) {
                    nFormRule.matchingRules = getMatchingRules(nFormRule.key)
                }
            }
            executableRulesList.addAll(nFormRule.matchingRules.asIterable())
        }
        rulesHandler.executableRulesList = executableRulesList
    }

    fun registerSubjects(subjects: List<String>, viewProperty: NFormViewProperty) {
        subjects.forEach { subject ->

            val (key, dataType) = Utils.extractKeyValue(subject)
            setDefaultFact(key, dataType)

            if (!subjectsRegistry.containsKey(key)) {
                subjectsRegistry[key] = hashSetOf()
            }
            subjectsRegistry[key]?.add(
                NFormRule(
                    viewProperty.name,
                    hashSetOf()
                )
            )
        }
    }

    fun viewHasVisibilityRule(viewProperty: NFormViewProperty): Boolean {
        val hasVisibilityRule =
            allRules?.map { it.name }?.contains("${viewProperty.name}${Constants.RuleActions.VISIBILITY}")
        return hasVisibilityRule != null && hasVisibilityRule
    }

    private fun setDefaultFact(key: String, dataType: String) {
        when (dataType.toUpperCase(Locale.getDefault())) {
            DataTypes.BOOL.name -> facts.put(key, false)
            DataTypes.NUMBER.name -> facts.put(key, 0)
            DataTypes.TEXT.name -> facts.put(key, null)
            DataTypes.LIST.name -> facts.put(key, listOf<Any>())
            DataTypes.MAP.name -> facts.put(key, linkedMapOf<Any, Any>())
        }
    }

    private fun getMatchingRules(ruleName: String): Set<Rule> {
        if (allRules == null) {
            return hashSetOf()
        }
        return allRules?.filter { it.name.startsWith(ruleName) }!!.toSet()
    }

    enum class DataTypes {
        TEXT, BOOL, NUMBER, LIST, MAP
    }

    enum class RulesFileType {
        JSON, YAML
    }

    companion object {
        @Volatile
        private var instance: RulesFactory? = null

        val INSTANCE: RulesFactory
            get() = instance ?: synchronized(this) {
                RulesFactory().also {
                    instance = it
                }
            }
    }
}
